### ISSUE TITLE
feat: Check apk validity by referencing its certificate signature with known good one

### ISSUE THREAD

**Issue by qumuq-til** [Universal-Debloater-Alliance/universal-android-debloater-next-generation#336]
### Describe the feature you want

It may be a good idea to check for certificate signatures inside apks, stored in ./appname/META-INF/CERT.RSA and compare them with known good ones.

Certificate information can be printed out using the following command:
`openssl pkcs7 -in ./appname/META-INF/CERT.RSA -inform DER -print_certs -noout`

For example, on my device, running this check against com.android.egg (Android Easter Egg application) gives:
`subject=C = CN, ST = Shanghai, L = Xuhui, O = greatflag, OU = SW, CN = dazhi, emailAddress = dazhi@dazhi.sh.cn issuer=C = CN, ST = Shanghai, L = Xuhui, O = greatflag, OU = SW, CN = dazhi, emailAddress = dazhi@dazhi.sh.cn`

which clearly shows that a system app had been tampered with, because on a stock Android it should be:
`subject=C = US, ST = California, L = Mountain View, O = Google Inc., OU = Android, CN = Android issuer=C = US, ST = California, L = Mountain View, O = Google Inc., OU = Android, CN = Android`

Therefore I propose we compile a database of expected valid signatures, starting with Google apps and possibly extending to vendor apps later (need to discuss that), then perform those checks in uad-ng, marking re-assembled packages with an exclamation sign in the list.

I don't know how to print out signatures on the device itself, and whether `openssl` is enough, but I know of two other tools that can help, `keytool` and `apksigner`, none of which come preinstalled on devices.

### Acknowledgements

- [X] This issue is not a duplicate of an existing feature request.
- [X] I have chosen an appropriate title.
- [X] All requested information has been provided properly.

**Comment by Cosm-gedenke** [comment]
hello, I saw you self-assigned to this issue, does it mean you're going to solve it yourself or is it still open to a potential pull request fixing this? I have found a similar feature being implemented in a package manager for android made in java. it would be curious to see how this ought to be implemented in rust.

**Comment by Rudxain** [comment]
I'll quote myself from a chat:
> IMO, it's more about verifying "provenance" than security. It's helpful to know if a different APK with the same pack-name (ID) is installed, which helps explain weird inconsistencies between devices and OEMs. This could serve as a warning to the user that description and `removal` level may not match the actual app that's installed
>
> IIRC, Android (by default) doesn't allow replacing system apps with different signatures. So the requested feature doesn't tell users if they have malware installed, only that the OEM is abusing the namespace 😅

**Comment by qumuq-til** [comment]
It might very well be that they're simply compiling the whole thing from source including those google parts, and they have to sign it somehow, and obviously they can't sign it with google keys. I wonder if there is really a way for a vendor to package those google-provided apks instead. But it's bad either way, because we know that vendors can publish whatever functionality in those names. Then of course even if those apks were not tampered with, nothing precludes a vendor from installing system services that have complete control over the entire system, so that fact must also be noted in UAD documentation somewhere.

**Comment by Cosm-gedenke** [comment]
Ill go ahead and add that fact in documentations in my PR request, once it is done. You're in fact correct at the end of the day any vendor can repackage any apk and use Google signatures such that we don't have any assurance of whether the app is safe or not. We're dependent on the vendors and Google supervision of such. Obviously that's not very nice because we know mega corporations love our safety /s
Signatures only reduce the attack surface at the app level, and even then only partially.
Either way it's a step to ensure more safety, the only other step forward after is to request some database at runtime to check vulnerabilities in apks(which is not feasible, and a headache)
